File: aggregating-profiler/src/main/java/ru/fix/aggregating/profiler/Identity.java
package ru.fix.aggregating.profiler;

import javax.annotation.Nonnull;
import java.util.*;
import java.util.stream.Collectors;

public class Identity {
    final String name;
    final HashMap<String, String> tags = new HashMap<>();

    public Identity(String name) {
        this.name = name;
    }

    /**
     * @param tags should not contains null values. It should have even size (key-value)
     * */
    public Identity(@Nonnull String name, @Nonnull String... tags) {
        if(Arrays.asList(tags).contains(null)) {
            throw new NullPointerException("tags must not contains nulls. tags = " + Arrays.toString(tags));
        }
        if (tags.length % 2 != 0) {
            throw new IllegalArgumentException("Invalid tags array size: " + tags.length + ". Expected even size.");
        }

        this.name = name;

        for (int i = 0; i < tags.length - 1; i += 2) {
            this.tags.put(tags[i], tags[i + 1]);
        }
    }

    /**
     * @param tags should not contains null values or keys
     * */
    public Identity(@Nonnull String name, @Nonnull Map<String, String> tags) {
        if (tags.containsKey(null) || tags.containsValue(null)) {
            throw new NullPointerException("tags must not contains null keys or values. tags = " + tags);
        }

        this.name = name;

        for (Map.Entry<String, String> tag : tags.entrySet()) {
            this.tags.put(tag.getKey(), tag.getValue());
        }
    }


    public Map<String, String> getTags() {
        return Collections.unmodifiableMap(this.tags);
    }

    public boolean hasTag(String tagName, String tagValue) {
        return tags.containsKey(tagName) && tags.get(tagName).equals(tagValue);
    }

    public String getName() {
        return name;
    }

    public String identityString() {
        if (tags.isEmpty()) {
            return name;
        } else {
            return name +
                    tags.entrySet()
                            .stream()
                            .map(entry -> entry.getKey() + "=" + entry.getValue())
                            .sorted()
                            .collect(Collectors.joining(",", "{", "}"));
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Identity identity = (Identity) o;
        return name.equals(identity.name) &&
                tags.equals(identity.tags);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, tags);
    }

    @Override
    public String toString() {
        return identityString();
    }
}

File: aggregating-profiler/src/main/kotlin/ru/fix/aggregating/profiler/Identity.kt
package ru.fix.aggregating.profiler

import java.util.*
import kotlin.collections.HashMap

class Identity {
    @JvmField
    val name: String
    val tags: Map<String, String>
        get() = Collections.unmodifiableMap(field)

    /**
     * @param tags should have even size (key-value)
     */
    constructor(name: String, vararg tags: String) {
        require(tags.size % 2 == 0) {
            "Invalid tags array size: ${tags.size}. Expected even size."
        }
        this.name = name
        this.tags = tags.asList().chunked(2).associate { (a, b) -> a to b }
    }

    constructor(name: String, tags: Map<String, String>) {
        this.name = name
        this.tags = HashMap(tags)
    }

    fun hasTag(tagName: String, tagValue: String): Boolean {
        return tags.containsKey(tagName) && tags[tagName] == tagValue
    }

    fun identityString(): String = if (tags.isEmpty()) {
        name
    } else {
        name + tags.entries.map { "${it.key}=${it.value}" }.sorted().joinToString(",", "{", "}")
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val identity = other as Identity
        return name == identity.name && tags == identity.tags
    }

    override fun hashCode(): Int {
        return Objects.hash(name, tags)
    }

    override fun toString(): String {
        return identityString()
    }
}